From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: CN MrLin <cnmrlin@users.noreply.github.com>
Date: Sat, 3 Oct 2026 02:00:00 +0800
Subject: [PATCH] Allow the Pro photo slider to switch physical lenses

Enable the combined zoom range for low-resolution Pro JPEG photos. Switch
physical cameras at lens boundaries and keep the slider open during restart.

---
diff --git a/smali/H4/C.smali b/smali/H4/C.smali
--- a/smali/H4/C.smali
+++ b/smali/H4/C.smali
@@ -4967,6 +4967,19 @@
     const/16 v0, 0xd1
 
     :cond_3
+    iget v4, v3, Lcom/android/camera/fragment/i;->mCurrentMode:I
+
+    invoke-static {v3, v1, v4}, Lcom/android/camera/compat/ProPhotoZoom;->keepSlider(LH4/C;II)Z
+
+    move-result v4
+
+    if-eqz v4, :check_optical_slider
+
+    move v4, v10
+
+    goto :goto_1
+
+    :check_optical_slider
     invoke-virtual {v3}, LH4/C;->Qd()Z
 
     move-result v4
@@ -5722,6 +5735,14 @@
     iget v5, v3, Lcom/android/camera/fragment/i;->mCurrentMode:I
 
     invoke-static {v5, v10}, Lcom/android/camera/data/data/i;->j1(IZ)Z
+
+    move-result v5
+
+    if-nez v5, :cond_21
+
+    iget v5, v3, Lcom/android/camera/fragment/i;->mCurrentMode:I
+
+    invoke-static {v5}, Lcom/android/camera/compat/ProPhotoZoom;->useCombinedRange(I)Z
 
     move-result v5
 
diff --git a/smali/h9/C.smali b/smali/h9/C.smali
--- a/smali/h9/C.smali
+++ b/smali/h9/C.smali
@@ -3,6 +3,12 @@
 .source "SourceFile"
 
 
+# instance fields
+.field public volatile proZoomSwitchPending:Z
+
+.field public proZoomSwitchCamera:I
+
+
 # virtual methods
 .method public final J7(IFF)Z
     .locals 3
@@ -58,6 +64,25 @@
     goto :goto_0
 
     :cond_2
+    const/4 v0, 0x3
+
+    if-ne p1, v0, :original_zoom_action
+
+    iget v0, p0, Lg9/h;->c:I
+
+    invoke-static {v0}, Lcom/android/camera/compat/ProPhotoZoom;->useCombinedRange(I)Z
+
+    move-result v0
+
+    if-eqz v0, :original_zoom_action
+
+    invoke-static {p0, p2, p3}, Lcom/android/camera/compat/ProPhotoZoom;->switchLens(Lh9/C;FF)Z
+
+    move-result v0
+
+    return v0
+
+    :original_zoom_action
     iget-object p0, p0, Lg9/h;->b:Ljava/lang/ref/WeakReference;
 
     invoke-virtual {p0}, Ljava/lang/ref/Reference;->get()Ljava/lang/Object;
diff --git a/smali/com/android/camera/compat/ProPhotoZoom.smali b/smali/com/android/camera/compat/ProPhotoZoom.smali
new file mode 100644
--- /dev/null
+++ b/smali/com/android/camera/compat/ProPhotoZoom.smali
@@ -0,0 +1,84 @@
+.class public final Lcom/android/camera/compat/ProPhotoZoom;
+.super Ljava/lang/Object;
+
+.method public static useCombinedRange(I)Z
+    .locals 3
+    const/4 v0, 0x0
+    const/16 v1, 0xa7
+    if-ne p0, v1, :done
+    invoke-static {}, Lj9/f;->z3()Z
+    move-result v1
+    if-nez v1, :done
+    invoke-static {}, Lcom/android/camera/data/data/m;->l0()Z
+    move-result v1
+    if-nez v1, :done
+    invoke-static {}, Lg2/a;->a()Lr2/i1;
+    move-result-object v1
+    const-class v2, Lr2/S;
+    invoke-virtual {v1, v2}, LWh/b;->x(Ljava/lang/Class;)Ljava/lang/Object;
+    move-result-object v1
+    check-cast v1, Lr2/S;
+    invoke-virtual {v1, p0}, Lr2/S;->getComponentValue(I)Ljava/lang/String;
+    move-result-object v1
+    const-string v2, "JPEG"
+    invoke-virtual {v2, v1}, Ljava/lang/String;->equals(Ljava/lang/Object;)Z
+    move-result v0
+    :done
+    return v0
+.end method
+
+.method public static keepSlider(LH4/C;II)Z
+    .locals 4
+    const/4 v0, 0x0
+    const/16 v1, 0x80
+    if-eq p1, v1, :check
+    const/4 v1, 0x2
+    if-ne p1, v1, :done
+    :check
+    invoke-static {p2}, Lcom/android/camera/compat/ProPhotoZoom;->useCombinedRange(I)Z
+    move-result v2
+    if-eqz v2, :done
+    iget-object v2, p0, LH4/C;->O:LI9/r;
+    instance-of v3, v2, LI9/e;
+    if-eqz v3, :done
+    iget-object v2, v2, LI9/w;->S0:Landroid/util/Range;
+    if-eqz v2, :done
+    invoke-static {p2}, Lcom/android/camera/data/data/i;->m(I)Lv2/z0;
+    move-result-object v1
+    iget-object v1, v1, Lv2/z0;->e:Landroid/util/Range;
+    invoke-virtual {v2, v1}, Landroid/util/Range;->equals(Ljava/lang/Object;)Z
+    move-result v0
+    :done
+    return v0
+.end method
+
+.method public static switchLens(Lh9/C;FF)Z
+    .locals 4
+    iget-object v0, p0, Lg9/h;->b:Ljava/lang/ref/WeakReference;
+    invoke-virtual {v0}, Ljava/lang/ref/Reference;->get()Ljava/lang/Object;
+    move-result-object v0
+    check-cast v0, Lcom/android/camera/module/W;
+    const/4 v2, 0x0
+    if-eqz v0, :done
+    invoke-interface {v0}, Lcom/android/camera/module/W;->getCameraManager()Lj6/k;
+    move-result-object v1
+    invoke-interface {v1}, Lj6/k;->V()Lj9/a;
+    move-result-object v1
+    if-eqz v1, :done
+    iget v1, v1, Lj9/a;->a:I
+    iget-boolean v3, p0, Lh9/C;->proZoomSwitchPending:Z
+    if-eqz v3, :switch
+    iget v3, p0, Lh9/C;->proZoomSwitchCamera:I
+    if-ne v1, v3, :switch
+    const/4 v2, 0x1
+    return v2
+    :switch
+    iput v1, p0, Lh9/C;->proZoomSwitchCamera:I
+    const/4 v3, 0x1
+    iput-boolean v3, p0, Lh9/C;->proZoomSwitchPending:Z
+    invoke-static {p1, p2, v2, v1, v0}, Lg9/h;->l8(FFIILcom/android/camera/module/W;)Z
+    move-result v2
+    iput-boolean v2, p0, Lh9/C;->proZoomSwitchPending:Z
+    :done
+    return v2
+.end method
